### ISSUE TITLE
Knowledge base document missing

### ISSUE THREAD

**Issue by datacrud8** [KoushikJit/almost-md-knowledge-base#1]
Hi,
thanks for sharing this project. 
Question 1:
In the file pages/api/updatedatabase.ts , you mentioned that **diagnosticstest2.PDF and labdiagnostics.TXT** file under the path ./documents but in the in github its missing.

Can you please share those two files.

async function handleUpload(indexname: string, namespace: string, res: NextApiResponse) {
    const loader = new DirectoryLoader('./documents',{
        '.pdf': (path: string) => new PDFLoader(path, {
            splitPages: false
        }),
        '.txt': (path: string) => new TextLoader(path)
    });

Question 2:
In the another repo : https://github.com/KoushikJit/almost-md/blob/main/app/api/medichatgemini/route.ts you mentioned below index and namespace in PinconeDB and vectorsize around 19492
const retrievals = await queryPineconeVectorStore(pinecone, **'medic', "ns1"**, query);

So Is this the  "diagnosticstest2.PDF "  used to generate the namespace "ns1" ?

Or how did you generate this namespace with vector size  19492 ?

Question 3: 
From the app ui , user uploading .jpg or pdf , system generating summary and chat Q&A - This will be generating summary and Q&A based on the Pinecone Knowledge base data only correct ? which we uploaded previously diagnosticstest2.PDF correct ?

Please clarify above details and share the **diagnosticstest2.PDF and labdiagnostics.TXT** files 

thank you.

**Comment by datacrud8** [comment]
Can you please provide the above requested  diagnosticstest2.PDF and labdiagnostics.TXT files

**Comment by KoushikJit** [comment]
Hi these can be any pdf or text files that contain medical diagnostics information. I used pdf version of the popular medical text book "Henry's Clinical Diagnosis". I am not sure if I can distribute that textbook. 
I may suggest you to find out popular medical textbooks on clinical diagnostics. Purchase the pdfs and throw them in the knowledge base. Please make sure you are within the limits of "fair use" if you are using the material without permission from the author or the publisher. Hope this helps.

**Comment by KoushikJit** [comment]
for question 2 

Yes the "diagnosticstest2.pdf" is the Henry's Clinical Diagnosis book, which has about 1500 pages, and was used to create the ns1 with 19492 vectors.

**Comment by KoushikJit** [comment]
Question 3: 
For the extraction of the details from the report, the pinecone knowledge base won't be used. The extraction is like an OCR operation implemented using Gemini's vision capability using some prompt engineering.

While asking the question in the Chat Q&A, the pinecone knowledge base will be queried to fetch relevant information. And these "retrievals" will be used to generate the Answer for the question asked.

**Comment by KoushikJit** [comment]
Hope these help. Apologies for my late response. I'll try to keep a close eye on this and get things clarified asap.

**Comment by datacrud8** [comment]
thank you for the clarification details. 
One more suggestion & questions , may be useful for you to expand this app as well.
This is namespace:
const retrievals = await queryPineconeVectorStore(pinecone, 'medic', "ns1", query);
and knowledge is : "Henry's Clinical Diagnosis"

For example, user may ask about dental or vision or any other health related Q&A right.
We can't accommodate all those information in the single PDF correct ?

In this scenario, Can we upload multiple knowledge PDF files ?
user may ask any health related  Q&A , retrievals get  the response from the correct namespace .

How to implement this scenario?
Can we able to mentioned multiple namespace based on the knowledge ?
const retrievals = await queryPineconeVectorStore(pinecone, **'medic', "ns1", "ns2",  "ns3", query);**

thank you.

**Comment by KoushikJit** [comment]
That's a great suggestion for expansion. Yes, right now each run of the app allows us to upload multiple books, but all go into one singe namespace. In the subsequent runs you can specify different namespace names like 'namespace_dental', 'namespace_vision', and upload multiple dentistry or vision books.

However, in the companion app that queries the namespaces to fetch relevant retrievals, the current implementation can only query one single namespace specified in advance, namely 'ns1'. I could argue that even if I throw all the dentistry, vision, general diagnostics in one single giant namespace, only the relevant retrievals will come up in the similarity search.

BUT a better approach in my opinion would be if we can add a step to first reflect which knowledgebase to query to fetch the relevant snippets - is it dental, vision or general diagnostics. And then query those knowledgebases specifically to fetch the retrievals and build the final prompt. 

I am even thinking about building a complete agentic application in that case, that handles such reflection and orchestration automatically.

Thanks a lot for the suggestion :)

**Comment by datacrud8** [comment]
Thank you for the details.
It was working fine three weeks before but now last 1 week the response is not getting displayed. Its showing only the user question alone in the meanwhile relevant info getting loaded . Please see the attached screen shot.  
Is that anything recently changed ? Can you please validate from your system and confirm?

Also one more question which  tested previously , Is that support more than  1 page pdf ? here the issue is I can upload multi page pdf but if ask question from the 2nd page of pdf it was not giving any answer, simply saying not found .


![Capture1123](https://github.com/user-attachments/assets/8dbef2a8-0f69-431f-ac64-4e0c98ba0b58)


there is no issue on the system log:

POST /api/medichatgemini 200 in 3325ms
HEYYYY --- reqBody::::::: {
  messages: [
    { role: 'user', content: 'how is the report?' },
    {
      role: 'user',
      content: 'Is that any concerns with this report? '
    }
  ],
  data: {
    reportData: '**Diabetes Profile Sample Report**\n' +
      '\n' +
      "The patient's fasting glucose level of 83 mg/dl falls within the normal range (70-99 mg/dl). However, the total cholesterol is elevated at 221 mg/dl (high; reference range: 100-200 mg/dl) and triglycerides are significantly elevated at 1315 mg/dl (high; reference range: <150 mg/dl). HDL cholesterol is low at 22.5 mg/dl (reference range: >50 mg/dl), while LDL cholesterol is 36 mg/dl, falling within the optimal range (<100 mg/dl). HbA1c is 3.8%, indicating no signs of diabetic glucose metabolism (reference range: 2.9-4.2%). Other measured values, including albumin, urea nitrogen, creatinine, and total protein, are within their respective reference ranges.\n"
  }
}
   
 POST /api/medichatgemini 200

**Comment by KoushikJit** [comment]
Hi thanks for describing it in detail.
I think the issue was with the Vercel ai sdk's two packages ai and @ai-sdk/google.
Once I uninstalled and reinstalled the packages on my project directory, everything is working fine. 
I am closing the issue for now. Feel free to raise any other issue in the future. :)

**Comment by KoushikJit** [comment]
I missed the other issue, hence reopening.
Can you see if the details from the page 2,3,4 etc of the pdf report are being extracted in the "Report" component of the app on the left side? After that please make sure that you have confirmed the report by pressing the "Looks good" button.

In my case I have tested with this attached sample pdf report that has 4 pages. And the details from pages2,3,4 are extracted in the report component, like this

```
**Kidney Panel; KFT, Serum**

Significantly elevated levels were observed for urea (106.00 mg/dL, reference range: 17.00-43.00 mg/dL), creatinine (5.85 mg/dL, reference range: 0.51-0.95 mg/dL), uric acid (7.10 mg/dL, reference range: 2.60-6.00 mg/dL), and alkaline phosphatase (ALP) (158.00 U/L, reference range: 30.00-120.00 U/L). Phosphorus was also high (4.71 mg/dL, reference range: 2.80-4.00 mg/dL).

**Complete Blood Count (CBC)**

Hemoglobin (8.50 g/dL, reference range: 11.50-15.00 g/dL), packed cell volume (PCV) (27.10%, reference range: 36.00-46.00%), and red blood cell (RBC) count (2.83 million/mm³, reference range: 3.80-4.80 million/mm³) were considerably low. Red cell distribution width (RDW) was slightly elevated at 18.70% (reference range: 11.50-14.50%).

**Urine Examination, Routine**

Protein was present at 2+ (100 mg/dL), with a positive leukocyte esterase.  Microscopy revealed 18-20 WBC/HPF and few epithelial cells.

```
 
After that asking questions about the urine report, which is on page 3 does yield a response. 
like this 
```

The urine report shows signs of a urinary tract infection and likely kidney damage. The presence of protein (2+) indicates proteinuria, which can be a sign of kidney dysfunction. The positive leukocyte esterase and the high number of white blood cells (18-20 WBC/HPF) strongly suggest an infection. The few epithelial cells seen on microscopy are nonspecific and could be due to the infection or normal shedding. Taken together with the significantly elevated urea and creatinine in the kidney panel, the urine findings point towards a compromised kidney function, possibly due to infection or other underlying kidney disease.

```

**Please follow these steps and see if this works**
1. Upload this attached pdf report
[Report_Status_Final_Results_Units.pdf](https://github.com/user-attachments/files/17632732/Report_Status_Final_Results_Units.pdf)

2. Wait for extraction. See Urine examination details from page 3 are extracted or not.
3. Confirm using the red button. See if the "Report Added" badge is green on the top right corner of the page.
4. Then ask question related to the Urine examination report, and see if you are getting a correct response.

**Comment by datacrud8** [comment]
thank you its loading all pages.
One more question please, what is the unique feature for this app? 
I used chatgpt and uploaded same pdf and asked same question. Chatgpt also providing the best answer.
so please provide little more detail about the unique difference from chatgpt?

**Comment by KoushikJit** [comment]
Well specialized use cases like medical diagnosis requires specialized knowledge about specific fields. That can be achieved by either fine-tuning the foundational models like Chat GPT or by providing relevant retrievals from a specialized knowledge base at the time of prompting. In this project I have implemented the latter. In the first part of my youtube video I have made a case for RAG, and why that might be more useful over fine-tuning. 

I agree that the knowledge base has only one book at the moment. And I did plan to create an extensive medical knowledge base by feeding it additional specialized studies or books from more specialized fields like oncology, cardiology etc. But I guess I never got around to doing that.  :)
